Implement update-source-git command

FrNecas opened the pull request:
<!-- notes for reviewers -->

Just a few notes about the problems I ran into while working on this and which directly influence the current state of the command:
 - the check for patch changes is very simple, just an extension check for now. Taking the current dist-git spec file and using its patches for the checks is not sufficient as the patches may have been removed by a commit. If we wanted to be really precise with this check, we would have to check-out to each commit in the revision range and parse the spec file and its patches there, though I am not sure if that's what we want, seems a bit too complicated. Moreover, changing checkout on the provided local dist-git doesn't seem like very user-friendly IMO.
 - GitPython manipulation with diffs is quite inconvenient, there is no way how to reconstruct the patch that can be applied directly from a `Diff` object. There were 2 approaches that came to my mind to solve this:
    1. Check-out dist-git repo to the given commit and copy the modified files over to source-git.
    2. Use raw git API to get the raw patch and then git-apply to apply the changes and then commit
  - I don't really like the idea of checkout, as mentioned above, so I ended up going with 2) while handling renames and removals separately.

<!-- Links to other issues or pull requests,
     for cross-repository links use: ‹namespace›/‹repository›#‹ID of issue›
       (‹namespace›/‹repository›!‹ID of PR› respectively)
-->

Fixes: #1414 

---

<!-- release notes for changelog/blog follow -->
A new `packit source-git update-source-git` command has been introduced for taking new changes from dist-git (specified by a revision range) to source-git. These may include any changes except source code, patches and `Version` tag changes in the spec file.

csomh commented:
Didn't read through the code yet, but a quick test failed to update the source-git repo.

Here is what I've done:
1. I've used a source-git repo for `acl`, I've pushed it [here](https://gitlab.com/csomh/acl-src-git). This was created using `packit source-git init` from [the last but least commit](https://src.fedoraproject.org/rpms/acl/c/0ca099649455b0d1ba5f8b246ac63496d5d761e2?branch=rawhide) of the rawhide branch from rpms/acl. The source-git repo was cloned to `src/acl`.
2. The dist-git repo was [rpms/acl](https://src.fedoraproject.org/rpms/acl), checked out at [the last commit](https://src.fedoraproject.org/rpms/acl/c/d5e8875d09943c95040d88b48ca2a2b80e06b9b3?branch=rawhide) (which is a mass-rebuild commit, bumping the release, and adding a new entry in the changelog). The dist-git repo is cloned to `rpms/acl`.
3. Both repositories were pristine.
4. Ran:
```
$ packit source-git update-source-git rpms/acl src/acl HEAD
2022-01-21 10:14:54.580 api.py            INFO   Release differs between dist-git and source-git (3 in dist-git and 2 in source-git). Versions are the same, continuing with the update.
2022-01-21 10:14:54.589 utils.py          ERROR  The sources file was modified in commit c3da26b6653bdb2d39a4459bc2d26469aba1b808 which is part of the provided range. Such operation is not supported.
```

The commit mentioned in the error message is [the first commit](https://src.fedoraproject.org/rpms/acl/c/c3da26b6653bdb2d39a4459bc2d26469aba1b808?branch=rawhide) of the rawhide branch.

Am I missing something here?

FrNecas commented:
@csomh Thanks for looking into it. I think the problem is the format of the revision range in your command. From the research PR (https://github.com/packit/research/pull/127 ) I understood that the revision range should behave like in `git log`. Using `git log HEAD` shows you the full commit history so running `update-source-git ... HEAD` actually goes through all the commits. It's a bit unintuitive and I too was a bit surprised when I found out. But this is the behaviour of `git log` and hence also of GitPython's `iter_commits`. If you want to sync only the latest commit, then `HEAD^..` or something similar is the way to go.

csomh commented:
I see. I got confused by the help text. I'll make a suggestion for improvement there, then.

csomh reviewed (commented):
This looks really nice! Thank you!

One more note: when initializing the a source-git repository `.distro/.gitignore` is currently [reset](https://github.com/packit/packit/blob/79cf1ab65a7266104f901994ba053d473e41d0d0/packit/source_git.py#L327) in order to behave correctly. But this means, that changes to `.gitignore` synced back from dist-git will (almost) always conflict. For now it would be nice to skip applying such changes of the `.gitignore` file in the dist-git root, and make a note of this in the help text.

csomh commented on packit/api.py:
```diff
@@ -211,6 +213,122 @@ def update_dist_git(
+                        f"which is not supported by this command."
+                    )
+
+        for i, commit in enumerate(commits):
```
Would be nice to have an info-log telling the number of commits that are going to be synced.

jpopelka commented on tests/integration/test_source_git_update_source_git.py:
```diff
@@ -0,0 +1,122 @@
+    """Checks that an error is thrown when sources in dist-git were modified."""
+    sourcegit, _ = sourcegit_and_remote
+    distgit, _ = distgit_and_remote
+    with open(distgit / "sources", "w") as sources:
+        sources.write("abcd")
```
```suggestion
    (distgit / "sources").write_text("abcd")
```

There are more places like this in the file

jpopelka commented on tests/integration/test_source_git_update_source_git.py:
```diff
@@ -0,0 +1,122 @@
+        file.write(content)
+    update_api.dg.commit("Add", "")
+    update_api.update_source_git("HEAD~1..")
+    with open(sourcegit / DISTRO_DIR / new_file, "r") as file:
+        assert file.read() == content
```
```suggestion
    assert (sourcegit / DISTRO_DIR / new_file).read_text() == content
```

jpopelka reviewed (commented):
Haven't tested but looks very good to me :1st_place_medal: :rocket: :star:

FrNecas commented on packit/api.py:
```diff
@@ -211,6 +213,122 @@ def update_dist_git(
+        """Update a source-git repo from a dist-git repo
+
+        Synchronizes the spec file and commits in the given revision range.
+        The Version tag must not have changed and the sources and patches
```
Good point, I removed the check for now. Or perhaps @TomasTomecek could provide more insight whether there was a reason to check for the versions that we may be missing.

FrNecas commented:
Thanks for the reviews, I learnt something new about Python! :champagne: Hopefully all of the things pointed out should be fixed now. The help message should now be a bit less confusing (or at least I hope so). I've also added some tests for the `.gitignore` changes and modified the update code to account for it correctly. I think that in some cases the `.gitignore` change may be applicable (perhaps the user manually changed it to work?) so I think that trying to apply and only skipping it if the `git-apply` fails should be fine.

csomh commented on packit/local_project.py:
```diff
@@ -578,8 +579,39 @@ def commit(
+    def get_commit_hunks(self, commit: git.Commit) -> List[str]:
+        """Get a list of hunks of the given commit."""
+        patch = self.git_repo.git.show(commit, format="", color="never")
+        hunk_start_re = re.compile(r"diff --git a/[\w/]* b/[\w/]*")
```
I think this regex will miss paths which have a `.` in them, like `tests/tests.yaml`. Updating one of the `get_commit_hunks` tests will reflect this, and something like this will resolve it:

```suggestion
        hunk_start_re = re.compile(r"diff --git a/.+ b/.+")
```

Due to this issue I see an `list index out of range` error when trying to convert a dist-git commit which modified `.gitignore` and `tests/tests.yaml` in the same `acl` repo I already mentioned.

FrNecas replied:
Good catch, fixed, thanks.

csomh reviewed (approved):
This seems to be ready to be merged. Thank you!
